daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/PartialTransaction.scala

```diff
@@ -225,7 +225,7 @@ private[speedy] case class PartialTransaction(
     context: PartialTransaction.Context,
     contractState: ContractStateMachine[NodeId]#State,
     actionNodeLocations: BackStack[Option[Location]],
-    disclosedContracts: ImmArray[DisclosedContract],
+    disclosedContracts: HashSet[DisclosedContract],
```

Comment:
We need to store all disclosed contracts "used" within a transaction. Their order is irrelevant and we have no desire to store duplicate disclosed contracts. So, modification of this field consists of only adding in new disclosed contracts and checking if the contract already exists.

Hence a `HashSet` is a better data structure to use here.

Reply:
You don’t necessarily. The alternative would be that you don’t track anything here and instead you call `inputContracts` on the resulting transaction afterwards to figure out what got used and filter by that. The advantage of that is that the code becomes simpler and you don’t have to worry about missing a contract (if inputContracts misses stuff we have biggerp roblems). Wdyt?

Reply:
Got you.

So, we pass in the Speedy disclosures to `PartialTransaction#init` (as prior to this changeset) - that provides the potential contract disclosure that we need to filter down the road.

When `PartialTransaction#finish` is called, we can then perform the filter (using `Transaction#inputContracts` and `PartialTransaction#disclosedContracts`) and use the result of this to define `PartialTransaction#Result.disclosedContracts`.

Reply:
Exactly